File: src/lib/__tests__/timelineViewport.test.ts
import { describe, it, expect } from "vitest";
import {
  HOUR_MS,
  MINUTE_MS,
  MIN_SPAN_MS,
  type TimeWindow,
  type Viewport,
  clampCenter,
  clampMsPerPx,
  pan,
  tickIntervalMs,
  ticks,
  timeToX,
  viewportForSpan,
  visibleRange,
  visibleSpanMs,
  xToTime,
  zoom,
} from "../timelineViewport";

const W = 1000; // track width px
// A 24h window ending at a round epoch.
const END = 1_700_000_000_000;
const WINDOW: TimeWindow = { startMs: END - 24 * HOUR_MS, endMs: END };

// A 1h-visible viewport centered 2h before the live edge.
const vp = (): Viewport => viewportForSpan(END - 2 * HOUR_MS, HOUR_MS, W, WINDOW);

describe("timelineViewport mapping", () => {
  it("is center-anchored: centerMs maps to the track middle and back", () => {
    const v = vp();
    expect(timeToX(v.centerMs, v, W)).toBeCloseTo(W / 2, 6);
    expect(xToTime(W / 2, v, W)).toBeCloseTo(v.centerMs, 6);
  });

  it("round-trips time ↔ x", () => {
    const v = vp();
    for (const x of [0, 250, 500, 750, 1000]) {
      expect(timeToX(xToTime(x, v, W), v, W)).toBeCloseTo(x, 6);
    }
  });

  it("visibleSpanMs reflects the zoom", () => {
    expect(visibleSpanMs(vp(), W)).toBeCloseTo(HOUR_MS, 3);
  });
});

describe("pan", () => {
  it("drag right goes back in time; drag left goes forward", () => {
    const v = vp();
    expect(pan(v, 100, W, WINDOW).centerMs).toBeLessThan(v.centerMs);
    expect(pan(v, -100, W, WINDOW).centerMs).toBeGreaterThan(v.centerMs);
  });

  it("cannot pan past the live edge (center stays ≤ end - halfSpan)", () => {
    const v = vp();
    const far = pan(v, -10_000, W, WINDOW); // huge drag toward 'now'
    const half = visibleSpanMs(far, W) / 2;
    expect(far.centerMs).toBeLessThanOrEqual(WINDOW.endMs - half + 1);
  });

  it("cannot pan past the window start", () => {
    const v = vp();
    const far = pan(v, 10_000, W, WINDOW);
    const half = visibleSpanMs(far, W) / 2;
    expect(far.centerMs).toBeGreaterThanOrEqual(WINDOW.startMs + half - 1);
  });
});

describe("zoom", () => {
  it("zooming in (factor > 1) shrinks the visible span about the center", () => {
    const v = vp();
    const z = zoom(v, 2, W, WINDOW);
    expect(z.centerMs).toBeCloseTo(v.centerMs, 6);
    expect(visibleSpanMs(z, W)).toBeLessThan(visibleSpanMs(v, W));
  });

  it("clamps zoom-in at MIN_SPAN and zoom-out at the window span", () => {
    const v = vp();
    const tightest = zoom(v, 1e6, W, WINDOW);
    expect(visibleSpanMs(tightest, W)).toBeCloseTo(MIN_SPAN_MS, 3);
    const widest = zoom(v, 1e-6, W, WINDOW);
    expect(visibleSpanMs(widest, W)).toBeCloseTo(24 * HOUR_MS, 3);
  });
});

describe("clamps with a degenerate window", () => {
  it("pins center to the midpoint when the window is narrower than the span", () => {
    const tiny: TimeWindow = { startMs: 0, endMs: 5 * MINUTE_MS };
    const c = clampCenter(0, W, (10 * MINUTE_MS) / W, tiny);
    expect(c).toBeCloseTo(2.5 * MINUTE_MS, 3);
  });

  it("never lets the visible span exceed a sub-10-min window", () => {
    const tiny: TimeWindow = { startMs: 0, endMs: 5 * MINUTE_MS };
    const mpp = clampMsPerPx(HOUR_MS / W, W, tiny);
    expect(mpp * W).toBeLessThanOrEqual(5 * MINUTE_MS + 1);
  });
});

describe("ticks", () => {
  it("picks a coarser interval as the span widens", () => {
    expect(tickIntervalMs(HOUR_MS)).toBeLessThan(tickIntervalMs(12 * HOUR_MS));
  });

  it("returns tick times within the visible range, on the interval grid", () => {
    const v = vp();
    const t = ticks(v, W);
    const { startMs, endMs } = visibleRange(v, W);
    const interval = tickIntervalMs(endMs - startMs);
    expect(t.length).toBeGreaterThan(0);
    for (const x of t) {
      expect(x).toBeGreaterThanOrEqual(startMs);
      expect(x).toBeLessThanOrEqual(endMs);
      expect(x % interval).toBe(0);
    }
  });
});

File: android/app/src/test/java/com/hawksnest/core/logic/TimelineViewportTest.kt
package com.hawksnest.core.logic

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/** Ports `timelineViewport.test.ts` 1:1. */
class TimelineViewportTest {

    private val w = 1000f
    private val end = 1_700_000_000_000L
    private val window = TimeWindow(end - 24 * HOUR_MS, end)

    // A 1h-visible viewport centered 2h before the live edge.
    private fun vp(): Viewport = viewportForSpan(end - 2 * HOUR_MS, HOUR_MS, w, window)

    private fun near(a: Double, b: Double, tol: Double = 1.0) =
        assertTrue(kotlin.math.abs(a - b) <= tol, "expected ~$b but was $a")

    @Test
    fun `center-anchored: centerMs maps to the track middle and back`() {
        val v = vp()
        near(timeToX(v.centerMs, v, w).toDouble(), (w / 2f).toDouble())
        assertEquals(v.centerMs, xToTime(w / 2f, v, w))
    }

    @Test
    fun `round-trips time to x`() {
        val v = vp()
        for (x in listOf(0f, 250f, 500f, 750f, 1000f)) {
            near(timeToX(xToTime(x, v, w), v, w).toDouble(), x.toDouble(), tol = 1.0)
        }
    }

    @Test
    fun `visibleSpanMs reflects the zoom`() {
        near(visibleSpanMs(vp(), w), HOUR_MS.toDouble(), tol = 1000.0)
    }

    @Test
    fun `drag right goes back in time; drag left goes forward`() {
        val v = vp()
        assertTrue(pan(v, 100f, w, window).centerMs < v.centerMs)
        assertTrue(pan(v, -100f, w, window).centerMs > v.centerMs)
    }

    @Test
    fun `cannot pan past the live edge`() {
        val v = vp()
        val far = pan(v, -10_000f, w, window)
        val half = (visibleSpanMs(far, w) / 2).toLong()
        assertTrue(far.centerMs <= window.endMs - half + 1)
    }

    @Test
    fun `cannot pan past the window start`() {
        val v = vp()
        val far = pan(v, 10_000f, w, window)
        val half = (visibleSpanMs(far, w) / 2).toLong()
        assertTrue(far.centerMs >= window.startMs + half - 1)
    }

    @Test
    fun `zooming in shrinks the visible span about the center`() {
        val v = vp()
        val z = zoom(v, 2f, w, window)
        assertEquals(v.centerMs, z.centerMs)
        assertTrue(visibleSpanMs(z, w) < visibleSpanMs(v, w))
    }

    @Test
    fun `clamps zoom-in at MIN_SPAN and zoom-out at the window span`() {
        val v = vp()
        val tightest = zoom(v, 1e6f, w, window)
        near(visibleSpanMs(tightest, w), MIN_SPAN_MS.toDouble(), tol = 1000.0)
        val widest = zoom(v, 1e-6f, w, window)
        near(visibleSpanMs(widest, w), (24 * HOUR_MS).toDouble(), tol = 1000.0)
    }

    @Test
    fun `pins center to the midpoint when the window is narrower than the span`() {
        val tiny = TimeWindow(0, 5 * MINUTE_MS)
        val c = clampCenter(0, w, (10 * MINUTE_MS).toDouble() / w, tiny)
        assertEquals(2L * MINUTE_MS + MINUTE_MS / 2, c) // 2.5 min
    }

    @Test
    fun `never lets the visible span exceed a sub-10-min window`() {
        val tiny = TimeWindow(0, 5 * MINUTE_MS)
        val mpp = clampMsPerPx(HOUR_MS.toDouble() / w, w, tiny)
        assertTrue(mpp * w <= 5 * MINUTE_MS + 1)
    }

    @Test
    fun `picks a coarser interval as the span widens`() {
        assertTrue(tickIntervalMs(HOUR_MS) < tickIntervalMs(12 * HOUR_MS))
    }

    @Test
    fun `returns tick times within the visible range on the interval grid`() {
        val v = vp()
        val t = ticks(v, w)
        val range = visibleRange(v, w)
        val interval = tickIntervalMs(range.endMs - range.startMs)
        assertTrue(t.isNotEmpty())
        for (x in t) {
            assertTrue(x in range.startMs..range.endMs)
            assertEquals(0L, x % interval)
        }
    }
}
